actionpack/lib/action_dispatch/routing/route_set.rb

```diff
@@ -292,7 +292,15 @@ def define_url_helper(mod, route, name, opts, route_key, url_strategy)
           mod.module_eval do
             define_method(name) do |*args|
               options = nil
-              options = args.pop if args.last.is_a? Hash
+
+              if args.last.is_a? Hash
+                options = args.pop 
+              end
+
+              if options.is_a? ActiveSupport::HashWithIndifferentAccess
+                options = options.symbolize_keys 
```

Comment:
I've concluded a small investigation on the topic. I'd like to share the results.

`ActionDispatch::Routing::RouteSet` has always worked only with symbol keys (I've checked starting with master and down to 3-2-stable).

To support string keys Rails provides two types of wrappers `ActionDispatch::Routing::UrlFor` and `ActionDispatch::Routing::RoutesProxy`. Both of which symbolize keys:
- https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/routing/url_for.rb#L156
- https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/routing/routes_proxy.rb#L30

`ActionDispatch::Routing::UrlFor` symbolizes keys when `#url_for` is invoked with hash options.

`ActionDispatch::Routing::RoutesProxy` is used to provide access to routes of engines from application and vice-versa. E.g. `main_app.url_for`. It symbolizes keys when named route helpers are invoked. 

In rails 3.2 named route helpers worked with string keys since they've invoked `ActionDispatch::Routing::UrlFor#url_for` underneath. Starting with rails 4.0 it's not true.

**Therefore we get inconsistent results when `main_app.route_path(options)` and `url_for(options)` symbolize options and `route_path(options)` does not.**

One more inconsistency that `Rails.application.routes.url_helpers.url_for` won't symbolize options, but `Rails.application.routes.url_helpers#url_for` will.

We can solve these inconsistencies by symbolizing only at `ActionDispatch::Routing::RouteSet#url_for` and `ActionDispatch::Routing::UrlHelper#call`. This way we'll support string keys while having the same amount of symbolize calls.

And even better if string keys support is dropped (as was announced), we can have only one optional symbolize_keys call in `ActionDispatch::Routing::RouteSet#url_for` when options are `HashWithIndifferentAccess`. 

What would you prefer?